### PR TITLE
feat(consensus): clique

### PR DESCRIPTION
**Pull request by rkrasiuk**
Implementation of Clique proof-of-authority consensus protocol.

This PR includes:
- Snapshot voting
- Header and block validation
- Clique header seal

This PR does not include:
- Forward downloader
- Clique instantiation
- Forckchoice tracker

### CODE REVIEW

**Comment by gakonst** [comment]
Where do we need this & how?

File: crates/consensus/src/clique/utils.rs [hunk 1]
```diff
@@ -0,0 +1,44 @@
+//! Utility methods for clique consensus.
+use std::collections::BTreeSet;
+
```
Comment (mattsse) [inline crates/consensus/src/clique/utils.rs:3]:
```suggestion
```

File: crates/net/ecies/src/algorithm.rs [hunk 2]
```diff
@@ -314,6 +314,7 @@ impl ECIES {
             &sigdata[..64],
             RecoveryId::from_i32(sigdata[64] as i32)?,
         )?;
+        println!("RECOVERY >> {:?}", RecoveryId::from_i32(sigdata[64] as i32)?);
```
Comment (mattsse) [inline crates/net/ecies/src/algorithm.rs:317]: rm?

**Review by mattsse — CHANGES_REQUESTED** [review]
this is a ton of stuff.

there are some changes that touch existing code base, which is fine imo.

not particularly familiar with clique.
test look good, is this also covered in hive?

**Comment by rkrasiuk** [comment]
hive doesn't test clique explicitly, but it's required to be able participate in rpc test suite

**Comment by gakonst** [comment]
The main way I'd be convinced into merging this as a large chunk would be if we actually had a Hive test which did not pass w/o this PR and passed with this PR. So let's try to get there?

**Comment by rkrasiuk** [comment]
@gakonst we did agree on it, just didn't finalize the scope

**Comment by rkrasiuk** [comment]
closing as stale and not planned for now

**Comment by mattsse** [comment]
as assumed we need this

**Pull request closed without merging**